firebase-firestore/src/main/java/com/google/firebase/firestore/core/OnlineQueryRunner.java

```diff
@@ -0,0 +1,70 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.core;
+
+import com.google.android.gms.tasks.OnCompleteListener;
+import com.google.android.gms.tasks.Task;
+import com.google.android.gms.tasks.TaskCompletionSource;
+import com.google.firebase.firestore.remote.RemoteStore;
+import com.google.firebase.firestore.util.AsyncQueue;
+import com.google.firebase.firestore.util.ExponentialBackoff;
+
+public class OnlineQueryRunner {
+  private AsyncQueue asyncQueue;
+  private RemoteStore remoteStore;
+
+  private int attemptsRemaining = 1;
+
+  private ExponentialBackoff backoff;
+  private TaskCompletionSource<Long> taskSource = new TaskCompletionSource<>();
+
+  public OnlineQueryRunner(AsyncQueue asyncQueue, RemoteStore remoteStore) {
+    backoff = new ExponentialBackoff(asyncQueue, AsyncQueue.TimerId.RETRY_TRANSACTION);
+    this.asyncQueue = asyncQueue;
+    this.remoteStore = remoteStore;
+  }
+
+  public Task<Long> runCountQuery(Query query) {
```

Comment:
Yeah, thanks for pointing that out, fixed it.

I agree with you in general..I'd still like to keep it because I am fairly confident we will have more features that would have to be first released this way (or maybe even only this way).

Reply:
OnlineQueryRunner still looks like it won't correctly handle multiple invocations of runCountQuery() because both attemptsRemaining and backoff are instance variables and should not be re-used for subsequent invocations of runCountQuery().

If OnlineQueryRunner really is intended to be re-used, then consider creating it as an instance variable of SyncEngine and re-using the instance instead of creating one-off instances of OnlineQueryRunner and using them exactly once.